test/data_base.cpp

```diff
@@ -1,401 +1,1148 @@
+   test_block_exists(instance, 1, *block1_ptr, settings.index_addresses, false);
+   test_block_exists(instance, 2, *block2_ptr, settings.index_addresses, false);
+   test_block_exists(instance, 3, *block3_ptr, settings.index_addresses, false);
+}
+
+BOOST_AUTO_TEST_CASE(data_base__pop_above_missing_forkpoint_hash___fails)
+{
+   create_directory(DIRECTORY);
+   database::settings settings;
+   settings.directory = DIRECTORY;
+   settings.index_addresses = false;
+   settings.flush_writes = false;
+   settings.file_growth_rate = 42;
+   settings.block_table_buckets = 42;
+   settings.transaction_table_buckets = 42;
+   settings.address_table_buckets = 42;
+  
+   data_base_accessor instance(settings); 
+   
+   static const auto bc_settings = bc::settings(bc::config::settings::mainnet);
+   const chain::block genesis = bc_settings.genesis_block;
+   BOOST_REQUIRE(instance.create(genesis));
+   
+   const auto block1 = read_block(MAINNET_BLOCK1);
+   auto out_headers = std::make_shared<header_const_ptr_list>(header_const_ptr_list{});
+
+   // setup ends
+
+   auto result = instance.pop_above(out_headers, config::checkpoint(block1.hash(), 0));
+#ifndef NDEBUG
+   BOOST_REQUIRE(!result);
+#else
+   BOOST_REQUIRE(result);
+#endif
+}
+
+#ifndef NDEBUG
+BOOST_AUTO_TEST_CASE(data_base__pop_above__wrong_forkpoint_height___fails)
+{
+   create_directory(DIRECTORY);
+   database::settings settings;
+   settings.directory = DIRECTORY;
+   settings.index_addresses = false;
+   settings.flush_writes = false;
+   settings.file_growth_rate = 42;
+   settings.block_table_buckets = 42;
+   settings.transaction_table_buckets = 42;
+   settings.address_table_buckets = 42;
+  
+   data_base_accessor instance(settings); 
+   
+   static const auto bc_settings = bc::settings(bc::config::settings::mainnet);
+   const chain::block genesis = bc_settings.genesis_block;
+   BOOST_REQUIRE(instance.create(genesis));
+   
+   const auto block1 = read_block(MAINNET_BLOCK1);
+   auto out_headers = std::make_shared<header_const_ptr_list>(header_const_ptr_list{});
+
+   // setup ends
+
+    BOOST_REQUIRE(!instance.pop_above(out_headers, config::checkpoint(genesis.hash(), 10)));
+}
+#endif
+
+BOOST_AUTO_TEST_CASE(data_base__pop_above__pop_zero___success)
+{
+   create_directory(DIRECTORY);
+   database::settings settings;
+   settings.directory = DIRECTORY;
+   settings.index_addresses = false;
+   settings.flush_writes = false;
+   settings.file_growth_rate = 42;
+   settings.block_table_buckets = 42;
+   settings.transaction_table_buckets = 42;
+   settings.address_table_buckets = 42;
+  
+   data_base_accessor instance(settings); 
+   
+   static const auto bc_settings = bc::settings(bc::config::settings::mainnet);
+   const chain::block genesis = bc_settings.genesis_block;
+   BOOST_REQUIRE(instance.create(genesis));
+   
+   const auto block1 = read_block(MAINNET_BLOCK1);
+   auto out_headers = std::make_shared<header_const_ptr_list>(header_const_ptr_list{});
```

Comment:
Is the empty list parameter required?

Reply:
Empty list parameter wasn't required. Removed.